src/nodes.coffee

```diff
@@ -1081,6 +1150,25 @@ exports.Value = class Value extends Base
 
     fragments
 
+  ast: ->
```

Comment:
`Value`s don't fit neatly into the base `ast()` method's "shape" since it could have different "top-level" `type`s (`MemberExpression` if it has properties, otherwise whatever its `@base` is). I'd had this as an override of `_toAst()`, overriding `ast()` is just as easy here

This is one of the more interesting transformations, we store `Value`s "linearly" as the base with its list of properties, whereas the AST expects nested `MemberExpressions`, eg `a.b.c` ends up as:
```
type: 'MemberExpression'
object:
  type: 'MemberExpression'
  object:
    type: 'Identifier'
    name: 'a'
  property:
    type: 'Identifier'
    name: 'b'
property:
  type: 'Identifier'
  name: 'c'
```

Reply:
First, I’ve been trying to always use `Object.assign` instead of `merge` and `extends`. Those helper functions aren’t needed anymore, and should really be refactored out of the codebase; but until that happens, I’m trying not to create more references to them.

Second, I hadn’t imagined `ast` as ever getting overridden. The AST you quote would imply to me that `Value`‘s `astProperties` method would return an object with two keys, `object` and `property`:
- The value of `object` is what we already have on the current instance of the class: `type: 'Identifier', name: 'a'`, from the class’s `base` child node.
- The value of `property` is a child node, which I presume is the class’s `properties` child. So `property: @properties.ast()`, with variations for when `properties` is an array, etc.

And so on. You don’t need to write code to drill down to all the children of the `Value`/`MemberExpression`, because you’re recursively calling `ast()` on the node class’s children. Or at least, that should be how it works.

Reply:
Object spread only came to Node in 8.6, part of ES2018: https://node.green/#ES2018-features-object-rest-spread-properties. We need to support back to Node 6, at least for now. I got a lot of pushback dropping Node 4, I don’t think we can drop Node 6 just yet.

I’ll have to look into what you’re saying later. In general I want to follow the object-oriented style of the node classes. Not sure what the best way to do that is, with also needing to deal with this weird shape.

Reply:
> I don’t think we can drop Node 6 just yet

My question is sort of the inverse, not whether we should consider dropping Node 6 but whether we should consider keeping the polyfill for object spread even though it's reached stage (4?) so that it's still easily usable against Node 6. In my mind it's worth entertaining that even just for the sake of being able to use object spread syntax in the Coffeescript codebase itself

> In general I want to follow the object-oriented style of the node classes

I don't think overriding `ast()` is violating object-oriented style. The base `ast()` is a template method, so subclasses that fit that "template" just override the individual subparts of the template as necessary. But for subclasses that don't fit the template, you just override the template method itself

In this instance, `Value`'s `ast()` implementation is encapsulating/taking responsibility for arranging its base value and properties into the right shape. It's still delegating responsibility in the typical way to its children (to generate their own AST representations), it's just then assembling those child ASTs in its `ast()` method (rather than `astProperties()`, which would be a very awkward fit since it's natural in this case to couple specifying the `type` and location data of the AST nodes the `Value` is creating with specifying its other properties)